core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -415,6 +415,8 @@ public void commit() {
     } catch (RuntimeException e) {
       if (!strictCleanup || e instanceof CleanableFailure) {
         Exceptions.suppressAndThrow(e, this::cleanAll);
+      } else if (strictCleanup) {
+        throw new CommitStateUnknownException(e);
```

Comment:
Let me see about adding some tests,  should be similar in principal to what I did in  https://github.com/apache/iceberg/pull/8397/files#diff-37f64b622d5c3e0e890d94773e84ef49ceb3c1e9b0c03da313db9266b0e75afbR2051